heat_pump_adoption_modelling/pipeline/supervised_model/hp_status_prediction.py

```diff
@@ -0,0 +1,523 @@
+# File: heat_pump_adoption_modelling/pipeline/supervised_model/hp_status_prediction.py
+"""
+Predict the HP status of a household using a supervised learning model.
+"""
+
+# ----------------------------------------------------------------------------------
+
+from heat_pump_adoption_modelling import PROJECT_DIR, get_yaml_config, Path
+from heat_pump_adoption_modelling.pipeline.preprocessing import feature_engineering
+from heat_pump_adoption_modelling.pipeline.supervised_model.utils import plotting_utils
+
+import pandas as pd
+import numpy as np
+from sklearn import svm
+
+import re
+
+from sklearn.pipeline import Pipeline
+from sklearn.impute import SimpleImputer
+from sklearn.preprocessing import MinMaxScaler
+from sklearn.decomposition import PCA
+
+from sklearn.linear_model import LogisticRegression
+from sklearn.linear_model import SGDClassifier
+
+from sklearn.model_selection import train_test_split
+from sklearn.model_selection import cross_validate
+from sklearn.metrics import accuracy_score
+from sklearn.model_selection import cross_val_score
+
+
+# Load config file
+config = get_yaml_config(
+    Path(str(PROJECT_DIR) + "/heat_pump_adoption_modelling/config/base.yaml")
+)
+
+FIG_PATH = PROJECT_DIR / config["SUPERVISED_MODEL_FIG_PATH"]
+SUPERVISED_MODEL_OUTPUT = str(PROJECT_DIR) + config["SUPERVISED_MODEL_OUTPUT"]
+
+
+prepr_pipeline = Pipeline(
+    [
+        ("imputer", SimpleImputer(missing_values=np.nan, strategy="median")),
+        ("min_max_scaler", MinMaxScaler()),
+        ("pca", PCA(n_components=0.9, random_state=42)),
+    ]
+)
+
+prepr_pipeline_no_pca = Pipeline(
+    [
+        ("imputer", SimpleImputer(missing_values=np.nan, strategy="median")),
+        ("min_max_scaler", MinMaxScaler()),
+    ]
+)
+
+
+best_params = {
+    "Linear Support Vector Classifier":  # { "alpha": 0.0001,"epsilon": 0.0001,
+    # "penalty": "l2","tol": 0.001,}
+    {
+        "alpha": 0.0001,
+        "epsilon": 0.0001,
+        "penalty": "elasticnet",
+        "tol": 0.0005,
+        "max_iter": 5000,
+    }
+}
+
+model_dict = {
+    "Logistic Regression": LogisticRegression(random_state=42),
+    "Linear Support Vector Classifier": SGDClassifier(random_state=42),
+    "Support Vector Classifier": svm.SVC(probability=True, random_state=42),
+}
+
+
+def get_HP_status_changes(df):
+    """For properties with multiple entries, get heat pump status at first
+    and last entry time.
+
+    Parameters
+    ----------
+    df: pandas.DataFrame
+        Dataframe for not de-duplicated EPC data.
+
+    Return
+    ---------
+    future_hp_status: pandas.DataFrame
+        Dataframe with future heat pump status."""
+
+    # Get properties with multiple entries
+    multiple_entry_epcs = df.loc[df["N_ENTRIES_BUILD_ID"] >= 2]
```

Comment:
Yes, I realised this can happen. I handle this during supervised model preprocessing but not in a general way. I'll take care of this in preprocessing.